### ISSUE TITLE
Code reloading in dev

### ISSUE THREAD

**Issue by Rush** [dry-rb/dry-system-rails#4]
Can I ask what is missing to make it happen?

**Comment by blelump** [comment]
Hi @Rush , last year I have tried to deal with it, but abandoned the idea as it was quite time consuming task. Since then [we're using Zeus][1] (we've also tried Shotgun [and there's Spring as well***], but Zeus is way faster) to make the _hot reloading_ within development mode and it works quite well.

What has to be done is that it needs to use `ActionDispatch::Reloader` features and in essence, reload all dependent constants tree that have been changed in the meantime (i.e. since last request). Behind the scenes, the `dry-system` utilizes a IoC container which include all the constants you've pointed to register (i.e. app directory) and that causes the underlying problem with hot reloading (because the constants registered within container do not get _refreshed_). 

To sum it up, don't be scared :smile: as  `dry-system` works well within Rails ecosystem. In case of tests, we're using RSpec and both unit and integration tests play well.

I'd also suggest to get familiar with [berg app][2], as it's good architectural utilization of `dry-system`, completely fine within Rails ecosystem as well.

[1]: https://github.com/dry-rb/dry-system-rails/issues/3#issuecomment-261288207
[2]: https://github.com/icelab/berg
*** https://github.com/rails/spring , but I am not sure about the internals of this library, needs further investigation.

**Comment by arielvalentin** [comment]
@blelump I believe that that this issue could be addressed by hooking into `ActionDispatch::Reloader` by using `require_dependency` instead of `require` when loading [component dependencies](https://github.com/dry-rb/dry-system/blob/master/lib/dry/system/container.rb#L561). This Railtie could [override](https://discourse.dry-rb.org/t/rails-development-and-dry-system/571/2) the system container implementation of `require_component` but it is marked as API private in `dry-system` so I am not sure how to best handle overriding that functionality without potentially running into compatibility issues in the future. 
Do you think that it would be unreasonable to override `require_component` to address this issue?

**Comment by arielvalentin** [comment]
This [PR](https://github.com/dry-rb/dry-system/pull/100) is the first step in making this work! 

There are a few more things to change in this Railtie but I am confident we can make the appropriate updates to this project that will allow us to support code reloading.

**Comment by arielvalentin** [comment]
I have opened a [draft PR](https://github.com/dry-rb/dry-system-rails/pull/8) to address this issue. I would prefer to wait on upstream features to be released in `dry-system` before merging into master.

**Comment by solnic** [comment]
This was addressed in `0.2.0` - please try it out.
